Refactor main part1

kappi43 opened the pull request:
* Change the main loop
* Split code into functions

kappi43 committed:
Refactoring main first part.

* Change the main round loop into an index
	based, to avoid some borrow checker related problems.
* Split code into functions.
* Move the all cards printing from players to main

kappi43 committed:
Fix clippy warnings

kappi43 committed:
Update main.rs

Fix clippy

mles2 commented on src/main.rs:
```diff
@@ -79,7 +58,37 @@ fn play_round(players: &mut Players, current_bet: &mut PokerCombination) {
 
-fn check_round_result(current_bet: &PokerCombination, all_cards: &Hand) -> bool {
+fn handle_call(players: &mut Players, current_bet: &mut PokerCombination, index: usize) {
+    let result = did_call_succeed(current_bet, &players.get_all_cards());
```
we need to find a way to add custom values to poker combinations - for example High Card King should beat HighCard Queen - right now we may stay here with general comparison, but more specific task for input/output handling and call comparison must be added

kappi43 replied:
You are right. I can add it to this MR, unless You think it would be to big?

mles2 replied:
May be big

mles2 commented on src/main.rs:
```diff
@@ -79,7 +58,37 @@ fn play_round(players: &mut Players, current_bet: &mut PokerCombination) {
+}
+
+fn handle_round_result(players: &mut Players, index: usize, result: bool) {
+    if result {
```
in general this "result" refers to call result and you can name it like this to not make confusion (result may be too general for such algorithms). Additionally I recommend to treat each turn of "call" in BlefGame to be the duel only between two players - actualPlayer and previousPlayer, when call is successful, then previous player is increasing number of cards, otherwise actualPlayer is increasing this number, we don't need to keep all players now at this level (as it seems to complicate modifications here)

kappi43 replied:
I have uploaded a proposal to fix this

mles2 commented on src/main.rs:
```diff
@@ -79,7 +58,37 @@ fn play_round(players: &mut Players, current_bet: &mut PokerCombination) {
+}
+
+fn did_call_succeed(current_bet: &PokerCombination, all_cards: &Hand) -> bool {
     all_cards.discover_combinations().contains(current_bet)
```
I belive that "call" refers to situation, when we want to check, if someone else is telling the truth. It should succeed when the other player is lying and we called them, and fail, when he is telling the truth.

mles2 replied:
And as I wrote, we need to add later some mechanism to keep values within PokerCombination (we can even use  Rust enums with smart comparator for this)

mles2 reviewed (dismissed):
You can merge it, but some additional work is needed here (we can do it later).

kappi43 committed:
A proposal for resolving one of the review comments

Moved handle round result to Players and changed the name of it to `increase_cards_to_deal`.
The method accepts a combination and two indexes.
Got rid of increase_previous_players_number_of_cards_to_deal.
Changed name of did_call_succeed and flipped the logic to align with review suggestions

kappi43 commented on src/main.rs:
```diff
@@ -51,26 +49,9 @@ fn play_round(players: &mut Players, current_bet: &mut PokerCombination) {
-                *current_bet = PokerCombination::None;
+                let len = players.len();
+                let previous_index = (current_index + len - 1) % len;
+                handle_call(players, current_bet, current_index, previous_index);
```
test

kappi43 replied:
I have entirely redone this part. Have a look :)

kappi43 committed:
Redone the main loop. Main refactor part 1

kappi43 committed:
Removed unused import

kappi43 committed:
Merge branch 'master' into refactor_main_part1

kappi43 committed:
post merge conflict fixes :)

mles2 commented on src/main.rs:
```diff
@@ -20,11 +18,11 @@ fn main() {
-    }
+    play_game(
+        Players::new(config.no_of_players),
+        PokerCombination::None,
```
Docelowo current_bet powinien się składać z kombinacji i wartości, ale to już inny issue raczej
